### PR TITLE
Add `notasklocal` effect

### PR DESCRIPTION
**Pull request by Keno**
Split out from #45272. This effect models the legality of moving code
between tasks. It is somewhat related to effect-free/consistent, but
only with respect to task-local state. As an example consider something
like:

```
global glob
function bar()
    @async (global glob = 1; some_other_code())
end
```

The newly created task is not effect-free, but it would be legal to inline
the assignment of `glob` into `bar` (as long it is inlined before the creation
of the task of `some_other_code` does not access `glob`). For comparison,
the following is neither `notls`, nor `effect_free`:

```
function bar()
    @async (task_local_storage()[:var] = 1; some_other_code())
end
```

The same implies to implicit task-local state such as the RNG state.

Implementation wise, there isn't a lot here, because the implicit
tainting by ccall is the correct conservative default. In the future,
we may want to annotate various ccalls as being permissible for
notls, but let's worry about that when we have a case that needs it.

### CODE REVIEW

File: base/compiler/abstractinterpretation.jl [hunk 1]
```diff
@@ -1998,7 +1998,8 @@ function abstract_eval_statement(interp::AbstractInterpreter, @nospecialize(e),
                 effects.effect_free ? ALWAYS_TRUE : TRISTATE_UNKNOWN,
                 effects.nothrow ? ALWAYS_TRUE : TRISTATE_UNKNOWN,
                 effects.terminates_globally ? ALWAYS_TRUE : TRISTATE_UNKNOWN,
-                #=nonoverlayed=#true
+                #=nonoverlayed=#true,
+                TRISTATE_UNKNOWN
```
Comment (aviatesk) [inline base/compiler/abstractinterpretation.jl:2002]:
```suggestion
                #=notls=#TRISTATE_UNKNOWN
```

File: base/compiler/types.jl [hunk 2]
```diff
@@ -76,41 +82,46 @@ function Effects(
     effect_free::TriState,
     nothrow::TriState,
     terminates::TriState,
-    nonoverlayed::Bool)
+    nonoverlayed::Bool,
+    notls::TriState)
```
[the remark this answers is not in the document]
Comment (Keno) [reply base/compiler/types.jl:86]: fair enough. The other effects do tend to be more verbose.

[the remark this answers is not in the document]
Comment (oscardssmith) [reply base/compiler/types.jl]: especially since TLS=Transport Layer Security.

[the remark this answers is not in the document]
Comment (halleysfifthinc) [reply base/compiler/types.jl]: That was how I initially read it.

**Pull request merged**